File: src/uk/co/ben_gibson/git/link/Url/Factory/BitbucketCloudUrlFactory.java
package uk.co.ben_gibson.git.link.Url.Factory;

import uk.co.ben_gibson.git.link.Git.*;
import uk.co.ben_gibson.git.link.UI.LineSelection;
import uk.co.ben_gibson.git.link.Url.Substitution.URLTemplateProcessor;

public class BitbucketCloudUrlFactory extends TemplatedUrlFactory
{

    public BitbucketCloudUrlFactory(URLTemplateProcessor urlTemplateProcessor)
    {
        super(
            urlTemplateProcessor,
            "{remote:url}/src/HEAD/{file:path}/{file:name}?at={branch}#lines-{line:start}:{line:end}",
            "{remote:url}/src/{commit}/{file:path}/{file:name}#lines-{line:start}:{line:end}",
            "{remote:url}/commits/{commit}"
        );
    }


    public boolean supports(RemoteHost host)
    {
        return host.isBitbucketCloud();
    }


    private String formatLineSelection(LineSelection lineSelection)
    {
        if (lineSelection.isMultiLineSelection()) {
            return String.format("lines-%d:%d", lineSelection.start(), lineSelection.end());
        }

        return String.format("lines-%d", lineSelection.start());
    }
}

File: src/main/kotlin/uk/co/ben_gibson/git/link/url/factory/BitbucketCloudUrlFactory.kt
package uk.co.ben_gibson.git.link.url.factory

import uk.co.ben_gibson.git.link.url.UrlOptions
import uk.co.ben_gibson.git.link.url.template.UrlTemplates
import uk.co.ben_gibson.url.Path
import uk.co.ben_gibson.url.URL

class BitbucketCloudUrlFactory : TemplatedUrlFactory(UrlTemplates.bitbucketCloud()) {
    override fun createUrl(baseUrl: URL, options: UrlOptions): URL {
        return super.createUrl(normaliseBaseUrl(baseUrl), options)
    }

    private fun normaliseBaseUrl(baseUrl: URL): URL {
        return baseUrl.copy(path = Path(baseUrl.path.toString().removePrefix("scm/")))
    }
}